jbpm-runtime-manager/src/main/java/org/jbpm/runtime/manager/impl/PerCaseRuntimeManager.java

```diff
@@ -116,66 +115,31 @@ public RuntimeEngine getRuntimeEngine(Context<?> context) {
         if (!(context instanceof ProcessInstanceIdContext || context instanceof CaseContext)) {
             logger.warn("ProcessInstanceIdContext or CaseContext shall be used when interacting with PerCase runtime manager");
         }
-        
-        if (engineInitEager) {
-            KieSession ksession = null;
-            Long ksessionId = null;
 
-            RuntimeEngine localRuntime = findLocalRuntime(contextId);
-            if (localRuntime != null) {
-                return localRuntime;
-            }
-            synchronized (this) {
-                ksessionId = mapper.findMapping(context, this.identifier);
-                if (ksessionId == null) {
-                    ksession = factory.newKieSession();
-                    ksessionId = ksession.getIdentifier();
-
-                    if (context instanceof CaseContext) {
-                        ksession.execute(new SaveMappingCommand(mapper, context, ksessionId, getIdentifier()));
-                    }
-                } else {
-                    ksession = factory.findKieSessionById(ksessionId);
-                }
-            }
+        RuntimeEngine localRuntime = findLocalRuntime(contextId);
+        if (localRuntime != null) {
+            logger.debug("Runtime engine found in local {} with context {}", localRuntime, contextId);
+            return localRuntime;
+        }
+
+        Long ksessionId = mapper.findMapping(context, this.getIdentifier());
+        if(ksessionId == null && mapper instanceof InternalMapper) {
+            // we fall back as process instance must be closed but not really case and kie session must be still there.
+            logger.debug("PerCaseRuntimeManager KieSessionId found fallingback for context {} and owner {}", context.getContextId(), getIdentifier());
+            ksessionId = ((InternalMapper) mapper).findLogMapping(context, identifier);
+        }
```

Comment:
Maybe we could put this into a separate method and defer this further in case we don't want to eagerly initialize the ksession? Before it was done in initializer.

Reply:
It has always been done during building of the runtime engine (not init) before getting the lock
https://github.com/kiegroup/jbpm/blob/7.44.0.Final/jbpm-runtime-manager/src/main/java/org/jbpm/runtime/manager/impl/PerCaseRuntimeManager.java#L187
So it is mandatory to compute the kie session id. Here we are not initializing anything. 
Actually lazy init performed this query twice:
https://github.com/kiegroup/jbpm/blob/7.44.0.Final/jbpm-runtime-manager/src/main/java/org/jbpm/runtime/manager/impl/PerCaseRuntimeManager.java#L669 (here two)
so we removed one query actually

Reply:
Actually same holds for eager initialization. Once it was called in init block, then just before getting the lock. So makes sense.